test/framework/deployment_helpers.go

```diff
@@ -179,51 +186,130 @@ type watchPodLogsInput struct {
 
-			// Create log metadata file.
-			logMetadataFile := filepath.Clean(path.Join(input.LogPath, input.DeploymentName, pod.Name, container.Name+"-log-metadata.json"))
-			Expect(os.MkdirAll(filepath.Dir(logMetadataFile), 0750)).To(Succeed())
+	selector, err := metav1.LabelSelectorAsSelector(input.LabelSelector)
+	Expect(err).NotTo(HaveOccurred())
 
-			metadata := logMetadata{
-				Job:       input.Namespace + "/" + input.DeploymentName,
-				Namespace: input.Namespace,
-				App:       input.DeploymentName,
-				Pod:       pod.Name,
-				Container: container.Name,
-				NodeName:  pod.Spec.NodeName,
-				Stream:    "stderr",
-			}
-			metadataBytes, err := json.Marshal(&metadata)
-			Expect(err).To(BeNil())
-			Expect(os.WriteFile(logMetadataFile, metadataBytes, 0600)).To(Succeed())
+	eventHandler := newWatchPodLogsEventHandler(ctx, input, selector)
 
-			// Watch each container's logs in a goroutine so we can stream them all concurrently.
-			go func(pod corev1.Pod, container corev1.Container) {
-				defer GinkgoRecover()
+	handlerRegistration, err := podInformer.AddEventHandler(eventHandler)
+	Expect(err).NotTo(HaveOccurred())
 
-				logFile := filepath.Clean(path.Join(input.LogPath, input.DeploymentName, pod.Name, container.Name+".log"))
-				Expect(os.MkdirAll(filepath.Dir(logFile), 0750)).To(Succeed())
+	go func() {
+		defer GinkgoRecover()
+		<-ctx.Done()
+		Expect(podInformer.RemoveEventHandler(handlerRegistration)).To(Succeed())
+	}()
+}
 
-				f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
-				Expect(err).NotTo(HaveOccurred())
-				defer f.Close()
+type watchPodLogsEventHandler struct {
+	//nolint:containedctx
+	ctx         context.Context
+	input       watchPodLogsInput
+	selector    labels.Selector
+	startedPods map[string]struct{}
+}
 
-				opts := &corev1.PodLogOptions{
-					Container: container.Name,
-					Follow:    true,
-				}
+func newWatchPodLogsEventHandler(ctx context.Context, input watchPodLogsInput, selector labels.Selector) cache.ResourceEventHandler {
+	return &watchPodLogsEventHandler{
+		ctx:         ctx,
+		input:       input,
+		selector:    selector,
+		startedPods: map[string]struct{}{},
+	}
+}
+
+func (eh *watchPodLogsEventHandler) OnAdd(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	eh.streamPodLogs(*pod)
+}
+
+func (eh *watchPodLogsEventHandler) OnUpdate(_, newObj interface{}) {
+	pod := newObj.(*corev1.Pod)
+	eh.streamPodLogs(*pod)
+}
 
-				podLogs, err := input.ClientSet.CoreV1().Pods(input.Namespace).GetLogs(pod.Name, opts).Stream(ctx)
+func (eh *watchPodLogsEventHandler) OnDelete(_ interface{}) {}
+
+func (eh *watchPodLogsEventHandler) streamPodLogs(pod corev1.Pod) {
+	if pod.GetNamespace() != eh.input.Namespace {
+		return
+	}
+	if !eh.selector.Matches(labels.Set(pod.GetLabels())) {
+		return
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		return
+	}
+	if _, ok := eh.startedPods[string(pod.GetUID())]; ok {
```

Comment:
I think the last non-nit finding: Let's use a sync.Map (the LoadOrStore func)

I think we have to be safe against concurrent calls here